tests/integration/long/upgrade/__init__.py

```diff
@@ -0,0 +1,179 @@
+# Copyright 2013-2017 DataStax, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from tests.integration import CCM_KWARGS, use_cluster, remove_cluster, MockLoggingHandler
+from tests.integration import setup_keyspace
+
+from cassandra.cluster import Cluster
+from cassandra import cluster
+
+from collections import namedtuple
+from functools import wraps
+from threading import Thread, Event
+from ccmlib.node import TimeoutError
+import time
+import logging
+
+try:
+    import unittest2 as unittest
+except ImportError:
+    import unittest  # noqa
+
+
+def setup_module():
+    remove_cluster()
+
+
+UPGRADE_CLUSTER = "upgrade_cluster"
+UpgradePath = namedtuple('UpgradePath', ('name', 'starting_version', 'upgrade_version', 'configuration_options'))
+
+
+class upgrade_paths(object):
+    """
+    Decorator used to specify the upgrade paths for a particular method
+    """
+    def __init__(self, paths):
+        self.paths = paths
+
+    def __call__(self, method):
+        @wraps(method)
+        def wrapper(*args, **kwargs):
+            for path in self.paths:
```

Comment:
It's not ideal, but `method(*args, **kwargs)` may call some methods which depend on `UPGRADE_CLUSTER` (like `start_upgrade`, which eventually calls `_upgrade_one_node`) , and it'd be hard to pass that argument to every test method so it can pass it along when it calls a base method.

Reply:
Yeah -- I generally prefer the argument-passing style, even if it means adding the param to lots of methods. I think it's easier to debug and read than setting values in one place and reading in another. But, this still works!